### ISSUE TITLE
System.InvalidCastException: Unable to cast object of type 'ModelParams' to type 'CreateModelParams'

### ISSUE THREAD

**Issue by lostllama** [JeremySkinner/FluentValidation#1066]
### System Details

- FluentValidation version: 8.1.3
- Web Framework version: ASP.NET Core 2.2
- DI Container: Autofac

### Issue Description

I'm submitting this issue in the hopes someone may have seen this issue before and have an idea what causes it. It happens sporadically within our Kubernetes environment. 3 pods running the same Docker image of our API, 1 pod will start consistently failing with this exception whereas sending the same request to the other 2 will result in success - same API endpoint, same validation code, same request object.

At first I wondered if maybe Autofac was getting confused, and that maybe I was inheriting `CreateModelParamsValidator` from `ModelParamsValidator` so it was instantiating the wrong one, but this of course isn't the case. I'm injecting `ModelParamsValidator` into `CreateModelParamsValidator` and copying its rules, but in the request where this is failing I'm using `ModelParamsValidator`.

#### Models

    public class ModelParams
    {
        public string Name { get; set; }

        public string PictureDataBase64 { get; set; }

        public DateTimeOffset? ModelDate { get; set; }

        public string ModelTitle { get; set; }
    }

----

    public class CreateModelParams : ModelParams, IModelCreationTokenParams
    {
        public FormId ModelRequestFormId { get; set; }
        public string ModelCreationToken { get; set; }
        public int? TimeZoneOffsetMinutes { get; set; }
    }

#### Validators

    public class ModelParamsValidator : AbstractValidator<ModelParams>
    {
        public ModelParamsValidator()
        {
            RuleSet("create", () =>
            {
                RuleFor(o => o.Name).NotNull().Length(1, 100);
                RuleFor(o => o.PictureDataBase64).NotEmpty();
                When(o => o.PictureDataBase64 != null, () =>
                {
                    RuleFor(o => o.PictureDataBase64).Must((u) => BeValidURL(u, false)).WithMessage("{PropertyName} is an invalid URL.");
                });
				
                When(o => o.ModelTitle != null, () => RuleFor(o => o.ModelTitle).MinimumLength(1).NotEmpty());
            });

            RuleSet("update", () =>
            {
                RuleFor(o => o.Name).NotNull().Length(1, 100);
                RuleFor(o => o.PictureDataBase64).NotEmpty();
                When(o => o.PictureDataBase64 != null, () =>
                {
                    RuleFor(o => o.PictureDataBase64).Must((u) => BeValidURL(u, true)).WithMessage("{PropertyName} is an invalid URL.");
                });
            });

            RuleSet("patch", () =>
            {
                When(o => o.Name != null, () => RuleFor(o => o.Name).NotNull().Length(1, 100));
                When(o => o.PictureDataBase64 != null, () =>
                {
                    RuleFor(o => o.PictureDataBase64).Must((u) => BeValidURL(u, false)).WithMessage("{PropertyName} is an invalid URL.");
                });
            });

            When(o => o.ModelTitle != null, () => { RuleFor(o => o.ModelTitle).Length(0, 100); });
            When(o => o.ModelTitle == null, () => RuleFor(o => o.ModelDate).Null());
            When(o => !string.IsNullOrWhiteSpace(o.ModelTitle) && o.ModelTitle.Length > 0, () => RuleFor(o => o.ModelDate.HasValue).Equal(true));
            When(o => o.ModelDate.HasValue && !string.IsNullOrWhiteSpace(o.ModelTitle), () =>
            {
                RuleFor(o => o.ModelTitle).NotEmpty();
                RuleFor(o => o.ModelDate).GreaterThan(DateTimeOffset.UtcNow.AddYears(-30));
            });

            When(o => !string.IsNullOrEmpty(o.PictureDataBase64) && o.PictureDataBase64 != "doNotChange", () =>
            {
                RuleFor(o => o.PictureDataBase64).Must(ImageDataUrl.IsValidLength).WithMessage("{PropertyName} is too big");
            });
        }

        private bool BeValidURL(string url, bool isUpdate)
        {
            if (string.IsNullOrEmpty(url))
            {
                return true;
            }

            if (isUpdate && url == "doNotChange") // doNotChange is for legacy app support
            {
                return true;
            }

            return ImageDataUrl.IsValid(url);
        }
    }

---

    public class CreateModelParamsValidator : AbstractValidator<CreateModelParams>
    {
        public CreateModelParamsValidator(ModelParamsValidator ModelParamsValidator)
        {
            When(o => o.ModelCreationToken == null, () =>
            {
                foreach (var rule in ModelParamsValidator)
                {
                    this.AddRule(rule);
                }
            });

            When(o => o.ModelCreationToken != null, () =>
            {
                RuleFor(o => o.Name).Null();
                RuleFor(o => o.ModelTitle).Null();
                RuleFor(o => o.ModelDate).Null();
                RuleFor(o => o.PictureDataBase64).Null();
                RuleFor(o => o.ModelCreationToken).NotEmpty();
            });

            When(o => o.TimeZoneOffsetMinutes.HasValue, () => RuleFor(o => o.TimeZoneOffsetMinutes).InclusiveBetween(-12 * 60, 14 * 60));
        }

        private bool BeValidURL(string url, bool isUpdate)
        {
            if (string.IsNullOrEmpty(url))
            {
                return true;
            }

            if (isUpdate && url == "doNotChange")
            {
                return true;
            }

            return ImageDataUrl.IsValid(url);
        }
    }

#### Validation code

    ModelParamsValidator _modelValidator; // resolved from Autofac
    ModelParams ModelParams = request.ModelParams;
    var validationContext = new ValidationContext<ModelParams>(ModelParams, new PropertyChain(), new RulesetValidatorSelector("default", "update"));
    var validationResult = await _modelValidator.ValidateAsync(validationContext, cancellationToken);

#### Request model

    var modelParams = new ModelParams
    {
        Name = "Test",
        PictureDataBase64 = "doNotChange",
        ModelDate = DateTime.Parse("2019-03-27T15:00:00.000Z"),
        ModelTitle = "test"
    };

#### Exception and stack trace

> System.InvalidCastException: Unable to cast object of type 'ModelParams' to type 'CreateModelParams'. at FluentValidation.Internal.ConditionBuilder`1.<>c__DisplayClass2_0.<When>g__Condition|0(ValidationContext context) in C:\Projects\FluentValidation\src\FluentValidation\Internal\ConditionBuilder.cs:line 40
> at FluentValidation.Internal.PropertyRule.ValidateAsync (C:\Projects\FluentValidation\src\FluentValidation\Internal\PropertyRule.cs:350)
> at FluentValidation.AbstractValidator`1.ValidateAsync (C:\Projects\FluentValidation\src\FluentValidation\AbstractValidator.cs:146)
> at Application.Services.UpdateModelHandler.Handle (UpdateModelRequest.cs:60)
> at MediatR.Pipeline.RequestPostProcessorBehavior`2.Handle
> at MediatR.Pipeline.RequestPreProcessorBehavior`2.Handle
> at Application.Controllers.ModelApiController.UpdateAsync (ModelApiController.cs:151)
> at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute
> at System.Threading.Tasks.ValueTask`1.get_Result
> at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync
> at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync
> at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow
> at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next
> at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextExceptionFilterAsync
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync
> at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync
> at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke
> at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke
> at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.InvokeCore
> at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext]

**Comment by JeremySkinner** [comment]
Hi. I think this is related to how you're copying rules from ModelParamsValidator:

```csharp
foreach (var rule in ModelParamsValidator)
{
	this.AddRule(rule);
}
```

Rules are intrinsicly tied to the validators that define them. They are not designed to be copied from one validator to another. They are inherently tied to the validator that defined them, and the type they were defined against. This is not supported and you will definitely run into strange issues if you do this. 

The error you're seeing is coming from one of the `When` conditions. Each condition block has a unique ID associated with it (that allows the result of the condition to be cached, so it's only executed one and not for each rule inside it). When you copy rules from one validator to another, the condition is brought across as well (in this case, all the conditions that you defined in the `ModelParamsValidator` end up being duplicated in the `CreateModelParamsValidator`. You will now end up with multiple conditions with the same ID, which means you'll end up with stale/incorrect results and type checks against the wrong type. So basically, don't do this :) 

If you want to re-use rules across validators, you have a couple of options. Firstly, you could use a base class with a generic constraint which holds the base rules:

```csharp
public abstract class ModelParamsValidatorBase<T> : AbstractValidator<T> where T:ModelParams {
   ctor() {
      //...shared rules...
   }
}

public class ModelParamsValidator : ModelParamsValidatorBase<ModelParams> {
}

public class CreateModelParamsValidator : ModelParamsValidatorBase<CreateModelParams> {
   ctor() {
      // Base rules automatically included.
     // Now include any rules specificd to CreateModelParams
     RuleFor(...)...
   }
 }
```

Alternatively, a clear solution is to compose the validators together with `SetValidator`, which is similar to what you were trying to do, but doesn't copy the rules - it just invokes the other validator directly:

```csharp
public CreateModelParamsValidator(ModelParamsValidator ModelParamsValidator) {
   RuleFor(x => x).SetValidator(ModelParamsValidator).When(x => x.ModelCreationToken == null);
}
```

Also I see in your example you often use the top-level `When` method even if there's only 1 rule that you're applying the condition to. If there's only a single rule that you need the condition, you should use the chained version of `When`:

```
// Instead of this:
When(o => o.TimeZoneOffsetMinutes.HasValue, () => RuleFor(o => o.TimeZoneOffsetMinutes).InclusiveBetween(-12 * 60, 14 * 60));

// Do this:
RuleFor(o => o.TimeZoneOffsetMinutes).InclusiveBetween(-12 * 60, 14 * 60).When(o => o.TimeZoneOffsetMinutes.HasValue);
```

For single-rule conditions, this is more performant as you don't need the overhead of dealing with the result cache, and is also much easier to read.

**Comment by lostllama** [comment]
So even though I'm not using the CreateModelParamsValidator, the fact that I'm including rules from ModelParamsValidator in it can lead to problems in using ModelParamsValidator?

Thanks for your input.

**Comment by JeremySkinner** [comment]
It's very hard to know for sure without a reproducable sample, but it seems very likely that's the problem. But yes you should definitely never copy rules from one validator to another.

Also I'd just double check you're running the same version of FluentValidation in all the instances of your application - there were some issues with the condition cache that were fixed in v8.1.2 so I just want to make sure you've not got an old version left over somewhere.

**Comment by lostllama** [comment]
OK. Understood. I'll make the appropriate adjustments. Thank you.
